src/ralph/converters/xapi/base.py

```diff
@@ -0,0 +1,126 @@
+    @post_dump
+    def get_xapi_statement(self, data, many):  # pylint: disable=unused-argument
+        """Returns the xApi Statement"""
+        return json.loads(Statement(**self.xapi_props).to_json())
+
+    def add_required_properties(self):
+        """Add the required actor, verb, object properies to xapi_props"""
+        self.xapi_props["actor"] = self.get_actor()
+        self.xapi_props["verb"] = Verb(**self.get_verb_kwargs())
+        self.xapi_props["object"] = self.get_object()
+
+    def add_optional_properties(self):
+        """Add the optional result, context, timestamp properies to xapi_props"""
+        self.add_optional_property("result", self.get_result())
+        self.add_optional_property("context", self.get_context())
+        self.add_optional_property("timestamp", self.get_timestamp())
+
+    def add_optional_property(self, name, value):
+        """Add an optional property by given name if the value is not None"""
+        if value is not None:
+            self.xapi_props[name] = value
+
+    def get_actor(self):
+        """Returns the actor property to xapi_props"""
+        name = self.edx_event["username"]
+        if not name:
+            name = "anonymous"
+        account = AgentAccount(name=name, home_page=HOME_PAGE)
+        return Agent(account=account)
+
+    def get_verb_kwargs(self):
+        """Returns tincan.Verb kwargs as a dict for xapi_props
+        should contain id and display key value pairs
+        """
+        raise NotImplementedError("get_verb_kwargs should be implemented")
+
+    def get_object(self):
+        """Returns the object property for xapi_props
+        it might be a tincan Activity, Agent, SubStatement or StatementRef
+        """
+        raise NotImplementedError("get_object should be implemented")
+
+    def get_result(self):  # pylint: disable=no-self-use
+        """Returns the result property for xapi_props"""
+        return None
+
+    def get_context(self):
+        """Returns the context property for xapi_props"""
+        return Context(
+            extensions=self.get_common_context_extensions(), platform=HOME_PAGE
+        )
+
+    def get_common_context_extensions(self):
+        """Returns the extensions property for xapi_props context"""
+        context = self.edx_event["context"]
+        event_ip = self.edx_event["ip"]
+        if event_ip:
+            event_ip = event_ip.exploded
+        extensions = {
+            XAPI_EXTENSION_ACCEPT_LANGUAGE: self.edx_event["accept_language"],
+            XAPI_EXTENSION_AGENT: self.edx_event["agent"],
+            XAPI_EXTENSION_COURSE_ID: context["course_id"],
+            XAPI_EXTENSION_COURSE_USER_TAGS: context.get("course_user_tags", {}),
+            XAPI_EXTENSION_HOST: self.edx_event["host"],
+            XAPI_EXTENSION_IP: event_ip,
+            XAPI_EXTENSION_ORG_ID: context["org_id"],
+            XAPI_EXTENSION_REFERER: self.edx_event["referer"],
+            XAPI_EXTENSION_USER_ID: context["user_id"],
```

Comment:
I haven't found a specific place where to put this information in the xAPI statement so I put it in the context.extentions

Although the [spec](https://github.com/adlnet/xAPI-Spec/blob/master/xAPI-Data.md#requirements-18) states that excessive usage of extensions is meaningless - It's preferable to keep those details somewhere in the xAPI statement (to be able to convert back to edx format for example). WDYT?

Reply:
It should be in the `actor` section.

Reply:
This make sens, but it seems the [xAPI spec](https://github.com/adlnet/xAPI-Spec/blob/1.0.3/xAPI-Data.md#41-extensions) has foreseen extensions only as part of `Activity Definitions`,  `context` property, or `result` property. [The actor property ](https://github.com/adlnet/xAPI-Spec/blob/1.0.3/xAPI-Data.md#242-actor) includes only optional `objectType`(string), optional `name` (string) and a [unique identifier ](https://github.com/adlnet/xAPI-Spec/blob/1.0.3/xAPI-Data.md#2423-inverse-functional-identifier) (mailto IRI or String or URI or [account object](https://github.com/adlnet/xAPI-Spec/blob/1.0.3/xAPI-Data.md#2424-account-object)) Where the account object only holds a `homePage` and `name` property.

Reply:
For Masha's project we decided to add edx's anonynous user key in the `actor.account.name` field.

Reply:
Anonymous users are referred as "student".

Reply:
Ah, ok, so we put edx `context.user_id` in `actor.account.name` instead of edx `username` ?
And what about the rest of the extensions above ?

Reply:
This is subject to changes in the coming weeks. We'll discuss about this IRL soon.